Added support for concurrency in NunitRunner

benpearson84 opened the pull request:
As discussed here : https://groups.google.com/forum/#!topic/codename-taurus/uYwW5g-0yCc my changes to the NunitRunner are enclosed.

NUnitRunner now supports a -concurrency argument on its command line, this spins up a thread per user and executes the test.  If -iterations is passed then it will divide this by the number of concurrent users and spread the load across them.

I need support to update the python code to add the concurrency value from the taurus yml file to the NunitRunner command line.

undera commented:
Cool!  I will look into this soon

undera commented:
Some comments:

- there is no need to divide iterations by concurrency, because existing logic of iterations is "for each VUser", so it is expected for `concurrency=10; iterations=50` to produce 500 total loops over script.
- on practice, launching all concurrent users at once creates unrealistic spike of stress. can we introduce `ramp_up` command-line option that will control time to gradually start VUs? It implies some logic for `--duration` and `--iterations` interpreting, but nothing too complicated.

benpearson84 commented:
Thanks, I'll take a look and update the PR.

undera commented:
Another important aspect of running concurrent test is how Taurus will get number of actual concurrent threads running, for each moment. The easiest is to write it into `WriteReport()` file line. But here's the trick, writing one file with multiple threads generating samples is tricky, because the tool must provide live results, as soon as they get created. I'm not sure this is implemented as expected in your current code. We'll need to figure this out.

benpearson84 commented:
@undera Ive made a couple of changes.

Removed the divide iterations by concurrent users so all virtual users run all iterations.

Added ramp_up commandline variable, which will delay the starting of a new thread so all virtual users are loaded during the ramp_up time specified.

In regards to feeding back test information to Taurus, does Taurus read from the report.ldjson file to get this information?  Your right its currently coded to hold all the data in memory then dump it into the report.ldjson at the end of the test run.  

I could look at flushing the objects to the report.ldjson more regulary?  That might work.  Any thoughts?

undera commented:
Regarding data writing, it's usually arranged like this:

- there is queue that is shared between all VU threads
- each VU thread puts sample data into queue
- there is dedicated thread that watches the queue and writes results into report file

It's a bit tricky to make sure queue gets flushed properly in case test is interrupted, but nothing too complex. I can point at example of implementing it in [Java](https://github.com/Blazemeter/taurus-java-helpers/tree/master/src/main/java/com/blazemeter/taurus/junit) and [Python](https://github.com/Blazemeter/apiritif/blob/master/apiritif/loadgen.py#L147)

undera commented:
Regarding ramp-up and duration control, I want to make sure that behavior will be same as all other executors.
The math of ramp-up and hold-for option is following:
- overall test duration is ramp_up+hold_for
- when test starts, limit timer starts ticking, so it is not limit for each thread, it is limit for whole process
- when duration is reached, test stops, regardless if any threads have stopped or still running, threads should be able to be interrupted/shutdown gracefully
- if duration limit is set and also iterations limit is set, each thread still stops on iteration limit, unless test is stopped by duration limit

benpearson84 commented:
Have changed the duration to include the ramp_up and all 4 points are met.

Overall duration is hold_for + ramp_up if included.

Each thread will end at the same time regardless of start time, because of above.  

If iterations complete before duration expires then the test will stop, and if duration is reached before iterations then the test will stop.

Working on the data writing thread, will get back to you.

benpearson84 commented:
Please review the latest changes.  A new writer thread is spun up to read from the queue and execute the WriteReport() method every 3 seconds.  Worker threads are running updating that list.  Ive tested and it feeds back test information to Taurus.

benpearson84 commented:
Modified csharp.py to include mapping of concurrency and ramp_up values to the cmdline object.
